[Perry Wong] iP

pwjj2000 opened the pull request:
## Duke 
> "Every 60 Seconds In Africa, A Minute Passes." – Jamal ([source](https://www.youtube.com/watch?v=7Zm1hPbmzPw))

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ VERY VERY VERY FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE** (for now)!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)
---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
        Application.launch(Main.class, args);
}
```

Commit:
Add Gradle support

pwjj2000 committed:
added Echo.java + amended Duke.java

pwjj2000 committed:
added javadocs for Level-1

pwjj2000 committed:
added List

pwjj2000 committed:
added Task.java

pwjj2000 committed:
added Task subclasses Deadline, Event, Todo

pwjj2000 committed:
amended run() to process in loops instead of recursion

pwjj2000 committed:
added inputs and expected outputs for runtest

pwjj2000 committed:
minor amendments to greeting message

pwjj2000 committed:
added DukeException

pwjj2000 committed:
updated test inputs/outputs

pwjj2000 committed:
added delete

pwjj2000 committed:
Add SavedTasks.java

pwjj2000 committed:
Store deadline dates as LocalDate instead of String

pwjj2000 committed:
Merge branch 'branch-Level-7'

pwjj2000 committed:
Merge branch 'branch-Level-8'

pwjj2000 committed:
Correct SavedTasks.java and minor amendments to List.java and Task.java

* Convert date format from file before instantiating Deadline/Event
* Write all data into file only at the end of file to account for mark/
unmark
* Remove print message in mark/unmark so that it will not be produced
during

pwjj2000 committed:
Correct List.java to assign the correct index upon initialising

pwjj2000 committed:
Restructure classes for more OOP

pwjj2000 committed:
Package into duke package and subpackages

pwjj2000 committed:
Merge branch 'add-gradle-support'

pwjj2000 committed:
Add Gradle and refactor code to run and build Duke with Gradle

pwjj2000 committed:
Add JUnit tests for Storage.java, TaskList.java

pwjj2000 committed:
Add Javadoc to more public methods in Command and its subclasses

pwjj2000 committed:
Standardize alignment in classes

pwjj2000 committed:
Add FindCommand.java

pwjj2000 committed:
Merge branch 'branch-A-JavaDoc'

pwjj2000 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ShowListCommand.java

pwjj2000 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/TaskList.java

pwjj2000 committed:
Add Javadoc for additional methods from Level-9

jonasgwt commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+    public static Command parse(String command) throws DukeException {
+        Scanner sc = new Scanner(command);
+
+        if (sc.hasNext("mark")) {
```
Would a `switch` be better here?

pwjj2000 replied:
Thanks for the suggestion! I have changed it to `switch` and it looks neater now.

jonasgwt commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,122 @@
+     * @param len New length of the list.
+     */
+    public void showAdd(Task t, int len) {
+        System.out.println("Got it. I've added this task:\n" + t.toString() + "\n" + "Now you have " + len
```
Good job on placing a line break to improve readability!

pwjj2000 replied:
Thanks!

jonasgwt commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+     */
+    public void printList() {
+        ArrayList<Task> list = this.taskArrayList;
+        list.forEach(t -> System.out.println(list.indexOf(t) + 1 + ". " + list.get(list.indexOf(t)).toString()));
```
Although this line is less than 120 char would it be better to still break it into two lines for readability?

pwjj2000 replied:
I'll consider breaking them down into smaller variables as well. Great catch!

jonasgwt commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,122 @@
+        System.out.println("Here are the matching tasks in your list:");
+    }
+
+    private String taskString(int len) {
```
Perhaps we should use a verb here for the method name

pwjj2000 replied:
Thanks! I will look for a more appropriate name. Any recommendations would be welcome as well!

Darren12345677 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+public class Parser {
+
+    //Exception Messages
+    private static final String NO_DESC_TODO = "☹ OOPS!!! Description of the todo can't be empty.";
+    private static final String NO_DESC_DEADLINE = "☹ OOPS!!! Description of the deadline can't be empty.";
+    private static final String NO_DESC_EVENT = "☹ OOPS!!! Description of the event can't be empty.";
+    private static final String NO_DATE_DEADLINE = "☹ OOPS!!! Date/time of the deadline can't be empty.";
+    private static final String NO_DATE_EVENT = "☹ OOPS!!! Date/time of the event can't be empty.";
+    private static final String NO_INDEX = "☹ OOPS!!! Index of the task to be marked/unmarked/deleted can't be empty.";
+    private static final String NO_TARGET = "☹ OOPS!!! Keyword of task to be found can't be empty.";
+    private static final String UNKNOWN_COMMAND = "☹ OOPS!!! I'm sorry, but I don't know what that means :-(";
```
I like how you name your static constants here!

pwjj2000 replied:
Thanks!

Darren12345677 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,122 @@
+            String description = sc.next();
+            if (!sc.hasNext()) {
+                throw new DukeException(NO_DATE_EVENT);
+            }
+            String at = sc.next();
+            sc.close();
+            return new AddCommand("event", description, at);
+        } else if (sc.hasNext("todo")) {
+            sc.useDelimiter("todo\\s*");
+            if (!sc.hasNext()) {
+                throw new DukeException(NO_DESC_TODO);
+            }
+            String description = sc.next();
+            sc.close();
+            return new AddCommand("todo", description, null);
+        } else if (sc.hasNext("delete")) {
+            sc.useDelimiter("delete\\s*");
+            if (!sc.hasNextInt()) {
+                throw new DukeException(NO_INDEX);
+            } else {
+                int i = sc.nextInt();
+                sc.close();
+                return new DeleteCommand(i);
+            }
+        } else if (sc.hasNext("find")) {
+            sc.useDelimiter("find\\s*");
+            if (!sc.hasNext()) {
+                throw new DukeException(NO_TARGET);
+            } else {
+                String s = sc.next();
+                sc.close();
+                return new FindCommand(s);
+            }
+        } else {
+            if (command.equals("bye")) {
+                return new ExitCommand();
+            } else if (command.equals("list")) {
+                return new ShowListCommand();
+            } else {
+                throw new DukeException(UNKNOWN_COMMAND);
```
Nested if-else statements could be difficult to read especially when it is long. Would switch statements be a better alternative here?

pwjj2000 replied:
Thanks for the suggestion! I have changed it to switch and it looks neater now.

Darren12345677 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,118 @@
+                    DateTimeFormatter alt = DateTimeFormatter.ofPattern("MMM d yyyy");
+                    DateTimeFormatter input = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+                    String dateAlt = input.format(alt.parse(date));
+                    Deadline d = new Deadline(description, dateAlt);
+                    Event e = new Event(description, dateAlt);
```
Could the naming be more meaningful? I think it would be better to give a more descriptive naming to avoid confusion later on.

pwjj2000 replied:
Assuming you are referring to `dateAlt`, I'll try to think of a more appropriate name. Thanks!

Darren12345677 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,41 @@
+
+    protected String description;
+
+    protected boolean isDone;
```
I like the way you name this boolean attribute!

pwjj2000 replied:
Thanks!

Darren12345677 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,61 @@
+public class Duke {
+
+    private Storage storage;
+    private TaskList tasks;
```
I like the way you name the collection attribute here!

pwjj2000 replied:
Thanks!

Darren12345677 reviewed (commented):
Looks good! 👍
Overall, I found your code easy to read for the most part except for a few places
where the nesting was too deep and long.

pwjj2000 committed:
Fulfill CheckStyle requirements

pwjj2000 committed:
Add GUI

pwjj2000 committed:
Change if-else statements to switch cases

pwjj2000 committed:
Merge branch 'branch-A-CheckStyle'

pwjj2000 committed:
Merge branch 'branch-Level-10'

pwjj2000 committed:
Amend minor CheckStyle errors

pwjj2000 committed:
Add assert statements to important assumptions in the code

There are some code that requires certain assumptions to run, e.g.:
* Length of task list
* Task is either a Todo, Deadline or Event

Adding assert statements allows us to ensure that there are no loopholes
in the code and prevents errors from happening.

Let's add these statements to important points of the code such as
during Task instantiation and list updating.

Having assert statements is preferable over not having them as they help
to find mistakes in the code much quicker

pwjj2000 committed:
Amend code to achieve Code Quality

Code has some readability issues such as:
* Complicated Boolean expressions
* Deep nesting

This causes difficulty in deciphering code for the reader.

Let's simplify complicated expressions and extract out conditionals to
avoid deep nesting.

Making the above amendments allow readers to understand the code quicker
and easier.

pwjj2000 committed:
Add more assert statements to Mark, Unmark, Delete Commands

pwjj2000 committed:
Merge pull request #1 from pwjj2000/branch-A-CodeQuality

Amend code to achieve Code Quality

pwjj2000 committed:
Merge branch 'master' of https://github.com/pwjj2000/ip

* 'master' of https://github.com/pwjj2000/ip:
  Amend code to achieve Code Quality

pwjj2000 committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	data/tasks.txt
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/command/UnmarkCommand.java

pwjj2000 committed:
Amend minor CheckStyle errors due to indentation

pwjj2000 committed:
Merge pull request #2 from pwjj2000/branch-A-Assertions

Add assert statements to important assumptions in the code

pwjj2000 committed:
Merge branch 'master' of https://github.com/pwjj2000/ip

* 'master' of https://github.com/pwjj2000/ip:
  Amend minor CheckStyle errors due to indentation
  Add more assert statements to Mark, Unmark, Delete Commands
  Add assert statements to important assumptions in the code

pwjj2000 committed:
Add new Command Search that is more flexible than Find

Find can only find fully matching words.

Sometimes user does not know the full word of what he is searching for.

Let's add Search to compensate for the above.

This adds more flexibility to find tasks even with partially matching
strings.

pwjj2000 committed:
Fix code quality issues

pwjj2000 committed:
Merge branch 'branch-C-BetterSearch'

pwjj2000 committed:
Add a User Guide

pwjj2000 committed:
Add a representative screenshot

pwjj2000 committed:
Update README

pwjj2000 committed:
Change implementation of save

Save is done upon calling of ExitCommand initially.

Let's change it to save every time there are changes to the list.

This allows the local file to change automatically.

pwjj2000 committed:
Change minor DukeException handling

pwjj2000 committed:
Change javafx dependencies

pwjj2000 committed:
Amend JUnit method names
